File: kirun-java/src/test/java/com/fincity/nocode/kirun/engine/function/system/date/IsBeforeTest.java
package com.fincity.nocode.kirun.engine.function.system.date;

import java.util.Map;

import org.junit.jupiter.api.Test;

import com.fincity.nocode.kirun.engine.namespaces.Namespaces;
import com.fincity.nocode.kirun.engine.repository.reactive.KIRunReactiveFunctionRepository;
import com.fincity.nocode.kirun.engine.repository.reactive.KIRunReactiveSchemaRepository;
import com.fincity.nocode.kirun.engine.runtime.reactive.ReactiveFunctionExecutionParameters;
import com.google.gson.JsonArray;
import com.google.gson.JsonPrimitive;

import reactor.test.StepVerifier;

class IsBeforeTest {

    DateFunctionRepository dfr = new DateFunctionRepository();

    ReactiveFunctionExecutionParameters rfep = new ReactiveFunctionExecutionParameters(
            new KIRunReactiveFunctionRepository(), new KIRunReactiveSchemaRepository());

    @Test
    void yearTest() {

        JsonArray arr = new JsonArray();

        arr.add("year");

        rfep.setArguments(Map.of("dateone", new JsonPrimitive("2023-10-31T17:14:21.798Z"), "datetwo",
                new JsonPrimitive("2023-10-31T17:14:20.789Z"), "unit", arr));

        StepVerifier.create(dfr.find(Namespaces.DATE, "IsBefore").flatMap(e -> e.execute(rfep)))
                .expectNextMatches(r -> !r.next().getResult().get("result").getAsBoolean())
                .verifyComplete();
    }

    @Test
    void yearTest2() {

        JsonArray arr = new JsonArray();

        arr.add("year");
        rfep.setArguments(Map.of("dateone", new JsonPrimitive("2023-10-31T17:14:21.798Z"), "datetwo",
                new JsonPrimitive("2025-10-31T17:14:20.789Z"), "unit", arr));

        StepVerifier.create(dfr.find(Namespaces.DATE, "IsBefore").flatMap(e -> e.execute(rfep)))
                .expectNextMatches(r -> r.next().getResult().get("result").getAsBoolean())
                .verifyComplete();
    }

    @Test
    void dayTest3() {

        JsonArray arr = new JsonArray();

        arr.add("year");
        arr.add("day");

        rfep.setArguments(Map.of("dateone", new JsonPrimitive("2023-10-31T17:14:21.798Z"), "datetwo",
                new JsonPrimitive("2000-10-31T17:14:20.789Z"), "unit", arr));

        StepVerifier.create(dfr.find(Namespaces.DATE, "IsBefore").flatMap(e -> e.execute(rfep)))
                .expectNextMatches(r -> !r.next().getResult().get("result").getAsBoolean())
                .verifyComplete();
    }

    @Test
    void monthTest() {

        JsonArray arr = new JsonArray();

        arr.add("year");
        arr.add("month");

        rfep.setArguments(Map.of("dateone", new JsonPrimitive("2023-10-31T17:14:21.798Z"), "datetwo",
                new JsonPrimitive("2025-09-30T17:14:20.789Z"), "unit", arr));

        StepVerifier.create(dfr.find(Namespaces.DATE, "IsBefore").flatMap(e -> e.execute(rfep)))
                .expectNextMatches(r -> !r.next().getResult().get("result").getAsBoolean())
                .verifyComplete();
    }

    @Test
    void monthTest2() {

        JsonArray arr = new JsonArray();

        arr.add("year");
        arr.add("month");

        rfep.setArguments(Map.of("dateone", new JsonPrimitive("2023-10-31T17:14:21.798Z"), "datetwo",
                new JsonPrimitive("2024-10-30T17:14:20.789Z"), "unit", arr));

        StepVerifier.create(dfr.find(Namespaces.DATE, "IsBefore").flatMap(e -> e.execute(rfep)))
                .expectNextMatches(r -> !r.next().getResult().get("result").getAsBoolean())
                .verifyComplete();
    }

    @Test
    void daytest1() {

        JsonArray arr = new JsonArray();

        arr.add("day");

        rfep.setArguments(Map.of("dateone", new JsonPrimitive("2023-10-03T17:14:21.798Z"), "datetwo",
                new JsonPrimitive("2023-10-05T17:14:20.789Z"), "unit", arr));

        StepVerifier.create(dfr.find(Namespaces.DATE, "IsBefore").flatMap(e -> e.execute(rfep)))
                .expectNextMatches(r -> r.next().getResult().get("result").getAsBoolean())
                .verifyComplete();
    }

    @Test
    void secondtest1() {

        JsonArray arr = new JsonArray();

        arr.add("day");
        arr.add("second");

        rfep.setArguments(Map.of("dateone", new JsonPrimitive("2023-10-05T17:14:20.788Z"), "datetwo",
                new JsonPrimitive("2023-10-05T17:14:20.789Z"), "unit", arr));

        StepVerifier.create(dfr.find(Namespaces.DATE, "IsBefore").flatMap(e -> e.execute(rfep)))
                .expectNextMatches(r -> !r.next().getResult().get("result").getAsBoolean())
                .verifyComplete();
    }
}

File: kirun-js/__tests__/engine/function/system/date/IsBeforeTest.ts
import { KIRunFunctionRepository, KIRunSchemaRepository, Namespaces } from '../../../../../src';
import { DateFunctionRepository } from '../../../../../src/engine/function/system/date/DateFunctionRepository';
import { GetTimeZoneOffset } from '../../../../../src/engine/function/system/date/GetTimeZoneOffset';
import { FunctionExecutionParameters } from '../../../../../src/engine/runtime/FunctionExecutionParameters';

const dateFunctionRepo = new DateFunctionRepository();

test('test1 IsBeforeFunction', async () => {
    let isBefore = await dateFunctionRepo.find(Namespaces.DATE, 'IsBefore');
    let fep: FunctionExecutionParameters = new FunctionExecutionParameters(
        new KIRunFunctionRepository(),
        new KIRunSchemaRepository(),
    );

    if (!isBefore) {
        throw new Error('Function not available');
    }

    let arr = [];
    arr.push('year');
    arr.push('day');

    fep.setArguments(
        new Map<string, any>([
            ['dateone', '2023-10-31T17:14:21.798Z'],
            ['datetwo', '2023-10-31T17:14:20.789Z'],
            ['unit', arr],
        ]),
    );

    expect((await isBefore.execute(fep)).allResults()[0].getResult().get('result')).toBe(true);
});

test('test2 IsBeforeFunction', async () => {
    let isBefore = await dateFunctionRepo.find(Namespaces.DATE, 'IsBefore');
    let fep: FunctionExecutionParameters = new FunctionExecutionParameters(
        new KIRunFunctionRepository(),
        new KIRunSchemaRepository(),
    );

    if (!isBefore) {
        throw new Error('Function not available');
    }

    let arr = [];
    arr.push('year');

    fep.setArguments(
        new Map<string, any>([
            ['dateone', '2023-10-31T17:14:21.798Z'],
            ['datetwo', '2000-10-31T17:14:20.789Z'],
            ['unit', arr],
        ]),
    );

    expect((await isBefore.execute(fep)).allResults()[0].getResult().get('result')).toBe(false);
});
